reactor-core/src/main/java/reactor/core/publisher/SinkOneMulticast.java

```diff
@@ -25,58 +25,51 @@
 
 final class SinkOneMulticast<O> extends SinkEmptyMulticast<O> implements InternalOneSink<O> {
 
+	@SuppressWarnings("rawtypes")
+	static final Inner[] TERMINATED_VALUE = new Inner[0];
+
+	static final int STATE_VALUE = 1;
+
 	@Nullable
 	O value;
 
 	@Override
-	public EmitResult tryEmitEmpty() {
-		return tryEmitValue(null);
+	boolean isTerminated(Inner<?>[] array) {
+		return array == TERMINATED_VALUE || super.isTerminated(array);
 	}
 
 	@Override
-	@SuppressWarnings("unchecked")
-	public EmitResult tryEmitError(Throwable cause) {
-		Objects.requireNonNull(cause, "onError cannot be null");
+	public EmitResult tryEmitValue(@Nullable O value) {
+		Inner<O>[] prevSubscribers = this.subscribers;
 
-		Inner<O>[] prevSubscribers = SUBSCRIBERS.getAndSet(this, TERMINATED);
-		if (prevSubscribers == TERMINATED) {
+		if (isTerminated(prevSubscribers)) {
 			return EmitResult.FAIL_TERMINATED;
 		}
 
-		error = cause;
-		value = null;
-
-		for (Inner<O> as : prevSubscribers) {
-			as.error(cause);
-		}
-		return EmitResult.OK;
-	}
+		this.value = value;
```

Comment:
More of a question for my own understanding: Does the stashing of this value for future subscribers rely on the fact that the 'safe' entry point to this sink will be wrapped in a serialized sink? In other words, is it OK for the below test to fail?

```java
@RepeatedTest(5)
void test() throws InterruptedException {
  final SinkOneMulticast < String > sink = new SinkOneMulticast < > ();
  final AtomicReference < String > successfullyEmittedElement = new AtomicReference < > ();
  final AtomicReference < String > firstSubscriptionOnNext = new AtomicReference < > ();
  final CountDownLatch firstSubscriptionLatch = new CountDownLatch(1);

  // first subscription before element is emitted through sink
  sink.asMono()
    .subscribeOn(Schedulers.boundedElastic())
    .subscribe(a -> {
      firstSubscriptionOnNext.compareAndSet(null, a);
      firstSubscriptionLatch.countDown();
    });

  // two competing tryEmitValue calls -- I assume this is where I break the spec?
  RaceTestUtils.race(
    () -> {
      if (sink.tryEmitValue("foo").isSuccess()) {
        if (!successfullyEmittedElement.compareAndSet(null, "foo")) {
          fail("Expected only a single successful emit");
        }
      }
    },
    () -> {
      if (sink.tryEmitValue("bar").isSuccess()) {
        if (!successfullyEmittedElement.compareAndSet(null, "bar")) {
          fail("Expected only a single successful emit");
        }
      }
    }
  );

  firstSubscriptionLatch.await();
  assertThat(firstSubscriptionOnNext.get()).isEqualTo(successfullyEmittedElement.get());

  // second subscription *after* sink is 'terminated'
  final String secondSubscriptionOnNext = sink.asMono().block();
  assertThat(secondSubscriptionOnNext).isEqualTo(successfullyEmittedElement.get());

}
```

I assume it is my responsibility in this case to not break the spec by racing two `tryEmitValue` calls, but just want to double check.

Reply:
Absolutely correct. It is only possible to get there if you use Sinks.unsafe() spec which implies more advanced capabilities but unsafe